Prompt for slot capacity

jostephd opened the pull request:
## Motivation / Problem

When I create a slot from the orders or routing restrictions window, I often find myself adjusting its capacity right after creating it. That requires opening the trains list, opening the dropdown there, opening the slots window, finding the new slot in the list, and finally clicking the gear button.

I mentioned this in passing on #848.

Creating slots of capacity greater than 1 seems to be a common pattern. The Signalling page on the wiki has several examples involving such slots: [Example 4: Bidirectionally signalled lines](https://github.com/JGRennison/OpenTTD-patches/wiki/Signalling#example-4-bidirectionally-signalled-lines) has two such slots, [Example 5: Diverting slow trains into a siding to allow fast trains behind to overtake](https://github.com/JGRennison/OpenTTD-patches/wiki/Signalling#example-5-diverting-slow-trains-into-a-siding-to-allow-fast-trains-behind-to-overtake) also has such a slot, and [Example 1: Using slots and pathfinder deny to hold trains in queuing sidings](https://github.com/JGRennison/OpenTTD-patches/wiki/Signalling#example-1-using-slots-and-pathfinder-deny-to-hold-trains-in-queuing-sidings) could conceivably use such slots if the player wanted to have more than one recycling train loading at a time.

By the way, the first two of these (bidirectional line and express-NW) both use slots of "infinite" capacity. That seems to be a common pattern in its own right, so all the more reason to make it easier to create such slots.

In contrast, [Example 3: Using conditional orders and slots for dynamic dispatch of trains](https://github.com/JGRennison/OpenTTD-patches/wiki/Signalling#example-3-using-conditional-orders-and-slots-for-dynamic-dispatch-of-trains) specifically uses slot capacity of 1 in order to use "fewer trains than there are loading stations to be serviced". This example wouldn't benefit from this PR.

## Description

When creating a slot via the orders, routing restrictions, or progsig windows, right after prompting for the slot's name, prompt for its capacity.

This way, if player intends to increase the slot's capacity from the default, there's zero extra clicks needed, and if a capacity of 1 is fine, the query window can be closed easily (with a single click or keypress, and if the setting to allow closing windows by right-clicking them is on, that's a single click on a big target).

## Limitations

Relies on the recent slots list. It seemed like the easiest solution, since the code needs to figure out the new slot's id somehow, and it's literally right there at the top of the ring buffer.

When creating a slot from the actual slots list window, its capacity isn't prompted for. Should it?

I was kinda expecting it to blow up on me when I tried calling ShowQueryString from OnQueryTextFinished, but it does actually seem to work. The call tree looks like this:

* something calls ShowQueryString for the slot name
* * calls OnQueryTextFinished
* * * calls CloseWindowByClass
* * * * calls OnQueryTextFinished(nullptr)
* * * calls ShowQueryString for the slot capacity
* * * * calls OnQueryTextFinished

The PR adds calls to `Command<CMD_...>::Post`. Elsewhere OnQueryTextFinished uses `DoCommandP<CMD_...>`. I'm not sure if this makes a difference.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

JGRennison commented:
I'm not averse to the principle of setting the capacity at the same time as creating the slot.
As written, this will not work in multiplayer, and doesn't handle any errors/failures in creating the slot.
From a UX point of view, closing a query window to have another immediately pop up seems a bit odd. A single window with two fields in it would seem better to me, the name and capacity could then be passed in the same command call, though admittedly this does increase the amount of work required a fair bit.
In terms of frequency of use, roughly three quarters of the slots in my current game have capacity 1. One quarter of slots seems plenty enough for this to be worthwhile.

I will look into this more after work.

jostephd commented:
Thank you very much for the review! I hadn't considered MP, but of course if there's another player around they might be creating slots in parallel, and CcCreateTraceRestrictSlot would record _that_ as the most recent slot (if it was created successfully), so the PR as written makes an invalid assumption.

Thank you also for the code change you've made. I've tested it and it looks like all the plumbing is done and now all that's left is to add a prompt for the slot's occupancy and put the desired value in data.max_occupancy.

As to a query window with two strings, I have good news and bad news. Good news first, I have a prototype extended QueryStringWindow that displays either one textbox or two, and passes both values to OnQueryTextFinished:

![2025-05-07-123715_359x89_scrot](https://github.com/user-attachments/assets/25f8f765-7cd2-477a-948e-424db4afd36c)

It's not mergeable yet, but most of what's missing is to replace hardcoded stuff (like the labels' StringIDs) by caller-provided values. However, and this is the bad news, I've run into a problem that I'm not sure know how to solve:

In an ordinary QueryStringWindow, pressing TAB toggles fast forward even while the query window is open. In a two-textbox QueryStringWindow I'd want to override that so TAB switches focus between the two textboxes, and I _have_ been able to get TAB to do that by overriding OnKeyPress and calling SetFocusedWidget...but TAB not only switches focus _but also toggles fast-forward_, and I haven't been able to disable that by overriding OnKeyPress or OnHotkey or both.

TAB toggles fast-forward in other windows that have two textboxes too, like the Available Trains window in unpatched JGRPP and the even the Rail Station Selection window in upstream OpenTTD (tested with Industrial Stations Renewal), and this can't be disabled by OnHotkey or OnKeyPress in those windows, either.

<details>

<summary>It's because TAB presses aren't handled by the OnHotkey framework but directly by the video drivers.</summary>

It took some time to find, but it turns out that TAB is handled directly by the various video drivers:

https://github.com/JGRennison/OpenTTD-patches/blob/d7a62842e61f8259e92cd965cc3e15aa513a948b/src/video/allegro_v.cpp#L469-L471
https://github.com/JGRennison/OpenTTD-patches/blob/d7a62842e61f8259e92cd965cc3e15aa513a948b/src/video/sdl2_v.cpp#L961-L963
https://github.com/JGRennison/OpenTTD-patches/blob/d7a62842e61f8259e92cd965cc3e15aa513a948b/src/video/win32_v.cpp#L943-L945

Notably, TAB is _not_ handled by binding WKC_TAB to the "Fast Forward" main toolbar button (the 0 stands for an invalid keycode):

https://github.com/JGRennison/OpenTTD-patches/blob/d7a62842e61f8259e92cd965cc3e15aa513a948b/src/toolbar_gui.cpp#L2364

Oh, and the "Fast Forward" toolbar button doesn't get lowered by [a Window::HandleButtonClick call as in other hotkeys' handler functions](https://github.com/JGRennison/OpenTTD-patches/blob/d7a62842e61f8259e92cd965cc3e15aa513a948b/src/toolbar_gui.cpp), but by [a specific check in MainToolbarWindow::OnRealtimeTick](https://github.com/JGRennison/OpenTTD-patches/blob/d7a62842e61f8259e92cd965cc3e15aa513a948b/src/toolbar_gui.cpp#L2327-L2328). Basically, every realtime tick the MainToolbarWindow goes "Hey, did the video driver just toggle fast forward without telling me?".

</details>

So, I'm not sure what to do now. To recap:

* We'd like to prompt for a slot name and a maximum occupancy
* For UX reasons, we'd like to prompt for both in a single query window
* In a query window with two textboxes, people _will_ try to use TAB to switch focus to the second textbox
* That will fast-forward their game for them when they didn't mean to do that
* Preventing that would require changing the video drivers
* Any such change to the video drivers would probably be applicable not only to JGRPP but also to upstream OpenTTD

How'd you like to proceed?

JGRennison commented:
No part of the game uses the tab key to move focus like that, even if it is a common pattern elsewhere.
I don't personally think that the current tab to fast forward behaviour is useful, but no doubt some players make use of it, and I doubt that upstream sees the current behaviour as a problem.
The only useful thing that the tab key is currently used for is tab-completion in the console window, which still does trigger the fast-forward behaviour if the game isn't paused.

Given that it's effectively impossible to play the game with only keyboard input, I don't think that not having a key to move focus like that would be a blocker for having a window with this layout/functionality. It is (in my opinion) better than using two windows or having to separately go through the slot window, whether there is such a key or not.

Perhaps it may be worth adding a separate setting to disable tab to fast forward in favour of something more useful (e.g. this sort of focus change behaviour)? Probably that should be a separate change/PR though.

jostephd commented:
Yeah, no other part of the game uses TAB to move focus like that, but on the other hand, we're talking about a text input window with two separate text input fields. I mean, I hope I'm not coming across as obstinate, but this window sounds (to me, anyway) kinda like the answer to "If you could add any one thing to OpenTTD to make people use the TAB key when playing, including even people who normally don't use the keyboard _at all_, what would you add?".

But anyway, I'm not trying to argue this issue to death. I'll implement a single query window with two strings, and leave TAB handling (both in the query window and outside of it) out of this PR entirely, as you said. I'll let you know when I have something ready for review.

I appreciate your help with this.

jostephd commented:
I've added a query window with two strings and rebased.

JGRennison commented:
Thanks for this.

I've pushed a few changes for some codestyle issues and to fix the dangling reference issue with the occupancy string parameters.

On the tab front, I had a bit of a rethink and added a change here: f0c616ab6e2ac60d6abcc8ec9b7228ba8e29b4bd. This flag could be set for the N>1 case.

In terms of the window layout, there should probably be some padding in front of the label, and aligning the two columns might look better?

JGRennison commented on src/programmable_signals_gui.cpp:
```diff
@@ -511,7 +511,12 @@ class ProgramWindow : public Window {
-	virtual void OnQueryTextFinished(std::optional<std::string> str) override
+	virtual void OnQueryTextFinished(std::optional<std::string> str) override final
+	{
+		OnQueryTextFinished(str, {});
```
Should these be std::nullopt?

jostephd replied:
I think it makes no difference to the generated code whether the parameter is {} or std::nullopt, and I don't find either one particularly easier to read over the other. So, for me, it's a six of one, half a dozen of the other situation.

jostephd commented:
Thank you very much for the codestyle and dangling reference fixes.

Also thanks for adding that tab flag. I'll add it for the N>1 case. To make that possible, I'll cherry-pick the tab commit into this branch (normally I'd rebase the branch, but you push commits to it too).

I'll check the padding and make sure the two-string window has as much padding as the one-string window in the jgrpp branch.

Aligning the two columns... could there be a problem in some non-English languages, or if one label is very long and the window is small? For non-pathological cases I guess it'll be nicer, so I'll implement that.

jostephd commented:
Added padding, alignment, and tab handling.

jostephd commented:
CI complains:

```
  Error: /home/runner/work/OpenTTD-patches/OpenTTD-patches/src/window.cpp:501:6: error: redefinition of ‘bool FocusedWindowSuppressesTabToFastForward()’
    501 | bool FocusedWindowSuppressesTabToFastForward()
        |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  /home/runner/work/OpenTTD-patches/OpenTTD-patches/src/window.cpp:492:6: note: ‘bool FocusedWindowSuppressesTabToFastForward()’ previously defined here
    492 | bool FocusedWindowSuppressesTabToFastForward()
        |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

But there's no definition on line 501 on the branch:

https://github.com/jostephd/OpenTTD-patches-JGRPP/blob/f976de0800a8838be595d0dcd5f7c88e652da32d/src/window.cpp#L488-L509

JGRennison commented:
Your branch seems to build fine on my machine. Probably the CI is building your PR after its been rebased, so that commit is being applied twice. It might be worth rebasing instead.
The alignment and tab handling sounds good. I will have a closer look at all this tomorrow.

jostephd commented:
I've rebased, which removed the cherry-picked copy of the tab video driver commit, and added a commit to fix padding in the one-string version of the query string window.

JGRennison commented:
Thanks for this :)
